### ISSUE TITLE
Notes about BNN implementation

### ISSUE THREAD

**Issue by stevengogogo** [stevengogogo/ECEN649_FinalProject#8]
## What is Bayesian neural network?

- The bayesian neural network is a standard neural network with weights in a probability distribution.

## What is the difference in training?

- The learning process depends on Bayes approach: $P(\theta  | D) = \frac{P(D|\theta) P(\theta)}{P(D)} \propto P(D|\theta) P(\theta)$
- The trick to avoid $P(D)$ is by using sampling methods.
- Sampling methods includes
  -  Metropolis-hasting Monte Carlo
  - Hamitonian Monte Carlo
  - Varitional Inference
  - Stochastic Langevin Descent

![Screen Shot 2022-11-28 at 11 20 19 AM](https://user-images.githubusercontent.com/29009898/204341010-d5c3096a-a379-438d-b4d1-8c1238cd0436.png)

## Implementation

For #5, I would recommend using Jax-Bayes that provides an example of BNN with regression problem. [The bnn code for regression](https://github.com/jamesvuc/jax-bayes/blob/master/examples/deep/nn_regression/mlp_regression_mcmc.py) has been tested on my laptop, it works as expected.

The idea is to apply this method to #5, since there dataset is organized, and can [be easily split into train-test set](https://github.com/sydney-machine-learning/Bayesianneuralnet_stockmarket/blob/6d24cf25115b6517e3099249bc657674f6b9b98f/code/nn.py#L71-L103). We could use Jax framework to train the model with [mcmc](https://github.com/jamesvuc/jax-bayes/blob/master/examples/deep/nn_regression/mlp_regression_mcmc.py) and [variational inference](https://github.com/jamesvuc/jax-bayes/blob/master/examples/deep/nn_regression/mlp_regression_var.py)

## Computational environment

- Since JAX can be easily installed on Colab TPU, we could transfer to Colab for massive computation

## Materias

- I highly recommend this article -- ["Hands-on Bayesian neural network"](https://arxiv.org/abs/2007.06823) that provides everything we need to know when implementing a BNN
  - This article provides [source code](https://github.com/french-paragon/BayesianNeuralNetwork-Tutorial-Metarepos) to implement BNN with Pytorch.
  - This tutorial provides classification example
- The Bayesian method series ([hierarchical linear model](https://twiecki.io/blog/2014/03/17/bayesian-glms-3/), [hierarchical BNN](https://twiecki.io/blog/2018/08/13/hierarchical_bayesian_neural_network/), [BNN](https://twiecki.io/blog/2016/06/01/bayesian-deep-learning/)) are another good resources on the statistical side.
  - This tutorials provides classification example 
  - The [source code of hierarchical BNN](https://github.com/twiecki/WhileMyMCMCGentlySamples/blob/master/content/downloads/notebooks/bayesian_neural_network_hierarchical.ipynb) based on PyMC with Theano backend.
- [JAX and BNN](https://neptune.ai/blog/bayesian-neural-networks-with-jax)

**Comment by stevengogogo** [comment]
Introduction to Bayesian interpretation of weight decay
---
https://www.youtube.com/watch?v=vEPQNwxd1Y4
